src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeMemory.cs

```diff
@@ -0,0 +1,132 @@
+        /// <remarks>This method is a thin wrapper over the C <c>malloc</c> API.</remarks>
+        [CLSCompliant(false)]
+        public static void* Alloc(nuint byteCount)
+        {
+            void* result = malloc(byteCount);
+
+            if (result == null)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            return result;
+        }
+
+        /// <summary>Allocates a block of memory of the specified size, in elements.</summary>
+        /// <param name="elementCount">The count, in elements, of the block to allocate.</param>
+        /// <param name="elementSize">The size, in bytes, of each element in the allocation.</param>
+        /// <returns>A pointer to the allocated block of memory.</returns>
+        /// <exception cref="OutOfMemoryException">Allocating <paramref name="elementCount" /> * <paramref name="elementSize" /> bytes of memory failed.</exception>
+        /// <remarks>This method is a thin wrapper over the C <c>calloc</c> API.</remarks>
+        [CLSCompliant(false)]
+        public static void* AllocZeroInitialized(nuint elementCount, nuint elementSize)
+        {
+            void* result = calloc(elementCount, elementSize);
+
+            if (result == null)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            return result;
+        }
+
+        /// <summary>Frees a block of memory.</summary>
+        /// <param name="ptr">A pointer to the block of memory that should be freed.</param>
+        /// <remarks>
+        ///    <para>This method does nothing if <paramref name="ptr" /> is <c>null</c>.</para>
+        ///    <para>This method is a thin wrapper over the C <c>free</c> API.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void Free(void* ptr)
+        {
+            free(ptr);
+        }
+
+        /// <summary>Reallocates a block of memory to be the specified size, in bytes.</summary>
+        /// <param name="ptr">The previously allocated block of memory.</param>
+        /// <param name="byteCount">The size, in bytes, of the reallocated block.</param>
+        /// <returns>A pointer to the reallocated block of memory.</returns>
+        /// <exception cref="OutOfMemoryException">Reallocating <paramref name="byteCount" /> of memory failed.</exception>
+        /// <remarks>
+        ///    <para>This method acts as <see cref="Alloc" /> if <paramref name="ptr" /> is <c>null</c>.</para>
+        ///    <para>This method acts as <see cref="Free" /> if <paramref name="byteCount" /> is <c>zero</c>.</para>
+        ///    <para>This method is a thin wrapper over the C <c>realloc</c> API.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void* Realloc(void* ptr, nuint byteCount)
+        {
+            void* result = realloc(ptr, byteCount);
+
+            if (result == null)
```

Comment:
This probably needs a fixup (or explicit decision to disallow) as `Realloc` behaves as `Free` for `byteCount == 0` and the result will be `null` in that scenario.

Reply:
The behavior for realloc with `byteCount == 0` is implementation defined in C11 https://www.cplusplus.com/reference/cstdlib/realloc/ . It means we should not depend on it. We should explicitly check for it and implement it as `malloc(0)`.

Reply:
> We should explicitly check for it and implement it as malloc(0).

As `return malloc(0);` or as `free(ptr); return null;`?

Reply:
`Marshal.ReAllocHGlobal` deals with this problem by bumping size 0 to 1. I would do the same here. I think always returning non-null from realloc is the more sensible behavior.

Note that `malloc(0)` has the same implementation-defined behavior. `Marshal.Alloc*` APIs convert it to `malloc(1)` for this reason to get consistent predictable behavior. I would do the same here.

Reply:
I'm almost thinking that bumping the size to `1` isn't the "right" behavior here. I think the more natural thing would be to check for `0` and return `nullptr` (treat it as OOM/invalid).
* I've implemented it as bumping to `1`, but wanted to raise the alternative of treating it as "invalid" (which is what some implementations do) and throwing `OOM`

Reply:
> I think the more natural thing would be to check for 0 and return nullptr

This is one of those things we can have end-less discussion about and there will be 50:50 split of the opinions. I think the most natural thing is to return non-null pointer for zero size.

According to https://www.cplusplus.com/reference/cstdlib/realloc/ : The spec for C90 was to return NULL. The spec for C99 is implementation defined. I guess that it was relaxed because many people did not consider returning NULL to be the sensible behavior and refused to standardize on it.